### PR TITLE
Preserve end-of-line comment for if statement without blocks

### LINKED ISSUE #10250: why the if comment is line breaked?
Is there any way to prevent line break?

**Prettier 2.2.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEMCGaDWAgpgLyYAUAlKQHybAC+AOlJFBtnvgEKkXUk6jFmw64CAYT5Va9ZlBYB6AFTKWmTMswABAA44ATjgC29AEYQIAGzg4oDTAEs0mALTY4HfUePrNilgAzAFdWGEdoDwxyZ2pgP0dAmLRqcSIZRUUnFxJczAAVA2C4WXyuQj84KzQSxPIAQljOAm4MrOdSPIAxHGqSwQKubkq+5vwJNuzOslCAEzhAxyg4WdKuCRZ5EAAaEAhdcPRkUEMDCAB3AAVDBDRkEF7znABPO92zIzB8OBgAZX0wEsAObIGBFOC7ODGMwreazAAydiBwRwQLgXQgBmMOBg4SgIJQOGCMAgOxAAAsYMYrAB1cmOeBoAFwX63BmOABuDOe90cQKgmLgZKWNQMMEuRiB2OQgV6NV2ACs0AAPbifb5-ExweFLIVIWV9RUq37AmwARWCEHgMrlEJA3lF9zMOBhVjJugMSxgNMcsxg5OQAA4AAy7D0QGo0oy6e6PF5vEAAR0t8Al+zuKA9njgBg5Qt2BjgycchYlqOl+ttuxqxkcNsNIDQprgFqterBxV2uDMPr9AaQACYu0ZHFZgRIIMYKyBPABWMnBGplMwZg3ykAc4oASSg81gvzAnoOhF3vxgzxs9ZqDAYQA)
```sh
--html-whitespace-sensitivity ignore
--parser babel
--prose-wrap always
--quote-props preserve
--no-semi
--single-quote
```

**Input:**
```jsx
const taskA = () => {}
const taskB = () => {}
const taskC = () => {}

/**
  * @param {boolean} is - test param
  */
function test(is) {
  if(is) taskA() // is === True => TaskA
  else if(!is) taskB() // is === False => TaskB
  else taskC() // is === undefined => TaskC
}

```

**Output:**
```jsx
const taskA = () => {}
const taskB = () => {}
const taskC = () => {}

/**
 * @param {boolean} is - test param
 */
function test(is) {
  if (is) taskA()
  // is === True => TaskA
  else if (!is) taskB()
  // is === False => TaskB
  else taskC() // is === undefined => TaskC
}

```

**Expected behavior:**
```jsx
const taskA = () => {}
const taskB = () => {}
const taskC = () => {}

/**
 * @param {boolean} is - test param
 */
function test(is) {
  if (is) taskA() // is === True => TaskA
  else if (!is) taskB() // is === False => TaskB
  else taskC() // is === undefined => TaskC
}

```

### LINKED ISSUE #12011: Preserve end of line comments on the same line
**Prettier 2.5.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBQENMC8RmADAJSYAmEAyhALZwwAWqUA5tmQNyYD0fTJHqNYmAIIAdKHAA2AZziYMOfEQKYAjBWp1GLNuwCiCuF14ChDUTEwAhaXMXKseQsQBMO2gyasOJorm-ILCNpgAwtIgADQgEAAOMKjQ8siguABOmRAA7gAKWQhpKLiyubgAnmlxAEaZuGAA1kw0uIwAMmxwyOhlinUNza0JjYbIMJkArnBxcPS1cJSUSx24HFO47HAAYhCZ9LgwyRzIILhTMBCxIMww9LIA6qzw8qNgcDTFqMkAbj+VM5geQ1NBQRSZGD5BrsQ69fqzEAAK3kAA8aIZZHAAIpTCDweGmOKjTIQs61XCLWQ3BKZNgwR6oSgsZAADhIxJyikeDQSZ1pcAhvx6cQAjnj4NDEiVzvIALQyJZLG6ZODi1Cq6FbOFIPpEkCKeioCbTRHyTE4iU9XUIuIwSmM5nMZAeO0NVCyQwRay4M6CgCsNymigAKpSSnqBiBfjMAJJQFawGhgOlJcQJmgwSpYwmKAC+eaAA)

```sh
--parser babel
```

**Input:**

```jsx
if (a === 0) doSomething(); // comment A
else if (a === 1) doSomethingElse(); // comment B
else if (a === 2) doSomethingElse(); // comment C

```

**Output:**

```jsx
if (a === 0) doSomething();
// comment A
else if (a === 1) doSomethingElse();
// comment B
else if (a === 2) doSomethingElse(); // comment C

```

**Expected behavior:**

Input is preserved as it is.

Comments at the end of line are put below the line with formatting, except the last statement.

**Related issues:**

This weird behavior around end of line comments have been reported before:

#807
#7155
#8708
#10250

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

Fixes #12011
Fixes #10250

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
**Prettier pr-12017**
[Playground link](https://deploy-preview-12017--prettier.netlify.app/playground/#N4Igxg9gdgLgprEAuEBLAZgAgBSSgE1RlWgEYBKTOADwAcAnOAZyZKlIG5MB6AKl+49BkALYiEMTAEEAOlDgAbJnEwYceQsWgAmSjQbNWOrnwFDMo8bEwAhOYuVU6jFmwDMJ-oO7CIYiZgAwiAANCAQtFpQTMigAIb09BAA7gAKCQgxKHEKyXEAnjFhAEb0cWAA1nAwAMpx4gAyqPLI6DnKJWWV1TW05c0A5sgw9ACucGFwIsVw+PizDXFQA6NxA3AAYhD0InEwxMvIIHGjMBChIAAWMCIKAOqXRMx9YHA1mUSoAG5E+UdgLAuzWU9BgqTKA12rXaExAACsmNQaoMFHAAIqjCDwaFKWF9eggo7FOIzBQXBjNGB3VD4GCXZAADgADGEGBBlHcyrQjgYQV84BcAI6Y+DgiJZY5MAC08lmswujGFqEY4LWUKQbVxYWUIlQwzGsNYy1RGKxAo1MLCMBJ1Np9KQ2itZVQCkGgT86pAzAArBdRsoACokrKajogL7jACSBAkNTA9FQkSkBBqMHyqJxygAvlmgA)

```sh
--parser babel
```

**Input:**

```jsx
if (condition1) expression1; /**/ // comment A
else if (condition2) expression2; /**/ // comment B
else expression3; /**/ // comment C
```

**Output:**

```jsx
if (condition1)
  expression1; /**/
  // comment A
else if (condition2)
  expression2; /**/
  // comment B
else expression3; /**/ // comment C

```

**Comment by sosukesuzuki** [comment]
@fisker I don't think such code is written in real world so we can ignore it.

**Comment by sosukesuzuki** [comment]
@fisker What do you think about this?

File: src/language-js/utils.js [hunk 1]
```diff
@@ -1315,6 +1315,12 @@ function isEnabledHackPipeline(options) {
   return Boolean(options.__isUsingHackPipeline);
 }
 
+/**
+ * This is used as a marker for dangling comments.
+ */
+const markerForIfWithoutBlockAndSameLineComment =
```
Comment (fisker) [inline src/language-js/utils.js:1321]: Can we use Symbol here?

**Review by fisker — APPROVED** [review]

File: src/language-js/utils.js (hunk 1 above)
Comment (sosukesuzuki) [reply src/language-js/utils.js]: 1c9bcef

**Pull request merged**